src/Storages/MergeTree/GinIndexStore.cpp

```diff
@@ -0,0 +1,447 @@
+    std::vector<std::pair<std::string_view, GinIndexPostingsBuilderPtr>> token_postings_list_pairs;
+    token_postings_list_pairs.reserve(current_postings.size());
+
+    for (const auto& [token, postings_list] : current_postings)
+    {
+        token_postings_list_pairs.push_back({std::string_view(token), postings_list});
+    }
+    std::sort(token_postings_list_pairs.begin(), token_postings_list_pairs.end(),
+                    [](const std::pair<std::string_view, GinIndexPostingsBuilderPtr>& a, const std::pair<std::string_view, GinIndexPostingsBuilderPtr>& b)
+                    {
+                        return a.first < b.first;
+                    });
+
+    ///write postings
+    std::vector<UInt64> encoding_lengths(current_postings.size(), 0);
+    size_t current_index = 0;
+
+    for (const auto& [token, postings_list] : token_postings_list_pairs)
+    {
+        auto encoding_length = postings_list->serialize(*postings_file_stream);
+
+        encoding_lengths[current_index++] = encoding_length;
+        current_segment.postings_start_offset += encoding_length;
+    }
+    ///write item dictionary
+    std::vector<UInt8> buffer;
+    WriteBufferFromVector<std::vector<UInt8>> write_buf(buffer);
+    FSTBuilder builder(write_buf);
+
+    UInt64 offset{0};
+    current_index = 0;
+    for (const auto& [token, postings_list] : token_postings_list_pairs)
+    {
+        String str_token{token};
+        builder.add(str_token, offset);
+        offset += encoding_lengths[current_index++];
+    }
+
+    builder.build();
+    write_buf.finalize();
+
+    /// Write FST size
+    writeVarUInt(buffer.size(), *term_dict_file_stream);
+    current_segment.item_dict_start_offset += getLengthOfVarUInt(buffer.size());
+
+    /// Write FST content
+    term_dict_file_stream->write(reinterpret_cast<char*>(buffer.data()), buffer.size());
+    current_segment.item_dict_start_offset += buffer.size();
+
+    current_size = 0;
+    current_postings.clear();
+    current_segment.segment_id = getNextSegmentID();
+
+    segment_file_stream->sync();
+    term_dict_file_stream->sync();
+    postings_file_stream->sync();
+}
+
+void GinIndexStoreReader::init_file_streams()
+{
+    String segment_file_name = store->getName() + ".gin_seg";
+    String item_dict_file_name = store->getName() + ".gin_dict";
+    String postings_file_name = store->getName() + ".gin_post";
+
+    segment_file_stream = store->storage->readFile(segment_file_name, {}, std::nullopt, std::nullopt);
+    term_dict_file_stream = store->storage->readFile(item_dict_file_name, {}, std::nullopt, std::nullopt);
+    postings_file_stream = store->storage->readFile(postings_file_name, {}, std::nullopt, std::nullopt);
+}
+void GinIndexStoreReader::readSegments()
+{
+    GinIndexSegments segments;
+
+    auto segment_num = store->getSegmentNum();
+    if (segment_num == 0)
+        return;
+
+    segments.assign(segment_num, {});
```

Comment:
Could use ctor [(3) or (4)](https://en.cppreference.com/w/cpp/container/vector/vector) of std::vector.

Reply:
Fixed.